<a href="https://colab.research.google.com/github/JPalaciosCh/ciclo_1_Programacion_101/blob/main/07_manipulaci_n_de_datos_con_pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://www.ctic.uni.edu.pe/wp-content/uploads/2022/04/588px-x-348px-web-1.png" alt="HTML5 Icon" width="900" height="350" >


# **Manipulación de Datos con Pandas**


**Objetivo**

El objetivo de este laboratorio es capacitar al estudiante con las herramientas y técnicas esenciales para la manipulación de datos utilizando la biblioteca Pandas. Al final del laboratorio, el estudiante será capaz de realizar operaciones básicas y avanzadas en conjuntos de datos, combinarlos, agruparlos y extraer información útil para el análisis.

**Contenido**

1. <a href="#item31">Introducción a Pandas</a>
2. <a href="#item31">Operaciones Básicas con DataFrames</a>
3. <a href="#item31">Combinación de Datasets</a>
4. <a href="#item31">Agregación y Agrupamiento</a>

</font>
</div>


---

# 1. Introducción a Pandas



## 1.1. ¿Qué es Pandas?

Pandas es una biblioteca de código abierto para el lenguaje de programación Python, que proporciona estructuras de datos y herramientas de análisis de datos de alto rendimiento y fáciles de usar. Es fundamental en el campo de la ciencia de datos y el análisis de datos.

**Características clave de Pandas:**

- Estructuras de datos flexibles: Series y DataFrames para datos unidimensionales y bidimensionales.
- Indexación potente: Permite acceso y manipulación eficientes de datos.
- Herramientas de limpieza y preparación de datos: Funciones para manejar datos faltantes, duplicados, y más.
- Integración con otras bibliotecas: Se complementa bien con NumPy, Matplotlib, y otras.



## 1.2. Estructuras de datos en Pandas

**Series:** Estructura unidimensional similar a una columna en una tabla.

In [ ]:
import numpy as np

In [ ]:
import pandas as pd

# Crear una Serie
s = pd.Series([1, 3, 5, 7, 9])
print(s)


0    1
1    3
2    5
3    7
4    9
dtype: int64


**DataFrame:** Estructura bidimensional, similar a una hoja de cálculo con filas y columnas.


In [ ]:
# Crear un DataFrame
data = {
    'Nombre': ['Ana', 'Luis', 'Carlos'],
    'Edad': [23, 45, 34],
    'Ciudad': ['Madrid', 'Barcelona', 'Valencia']
}

df = pd.DataFrame(data)
print(df)


In [ ]:
pd.DataFrame({'Age': [24, 30, 27, 21, 32, 38, 18],
              'Name': ['Adner', 'Brajan', 'Eder', 'Fernando', 'Hugo', 'Italo', 'Josue'],
              'Nota': np.random.randint(14, 20, 7)})

Age      Name  Nota
0   24     Adner    14
1   30    Brajan    16
2   27      Eder    15
3   21  Fernando    15
4   32      Hugo    19
5   38     Italo    14
6   18     Josue    17

## 1.3. Importación y exploración inicial de datos

Importar datos desde un archivo CSV:


In [ ]:
# archivo 'datos.csv'
df = pd.read_csv('titanic.csv')


**Exploración inicial:**

Ver las primeras filas:

In [ ]:
print(df.head())  # Muestra las primeras 5 filas

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


In [ ]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

Resumen de información:

In [ ]:
print(df.info())  # Información sobre tipos de datos y valores nulos

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Estadísticas descriptivas:

In [ ]:
print(df.describe())  # Estadísticas básicas de las columnas numéricas

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


In [ ]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [ ]:
df.shape

(891, 12)

In [ ]:
len(df)

891

In [ ]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [ ]:
df.columns.to_list()

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']

In [ ]:
df_female = df.loc[df.Sex == 'female']
df_female.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   314.000000  314.000000  314.000000  261.000000  314.000000   
mean    431.028662    0.742038    2.159236   27.915709    0.694268   
std     256.846324    0.438211    0.857290   14.110146    1.156520   
min       2.000000    0.000000    1.000000    0.750000    0.000000   
25%     231.750000    0.000000    1.000000   18.000000    0.000000   
50%     414.500000    1.000000    2.000000   27.000000    0.000000   
75%     641.250000    1.000000    3.000000   37.000000    1.000000   
max     889.000000    1.000000    3.000000   63.000000    8.000000   

            Parch        Fare  
count  314.000000  314.000000  
mean     0.649682   44.479818  
std      1.022846   57.997698  
min      0.000000    6.750000  
25%      0.000000   12.071875  
50%      0.000000   23.000000  
75%      1.000000   55.000000  
max      6.000000  512.329200

In [ ]:
df_male = df.loc[(df.Sex == 'female') & (df.Age < 16)]
df_male.describe()

PassengerId   Survived     Pclass        Age      SibSp      Parch  \
count    43.000000  43.000000  43.000000  43.000000  43.000000  43.000000   
mean    415.837209   0.651163   2.627907   7.209302   1.279070   1.186047   
std     280.745261   0.482243   0.618110   4.919570   1.368447   0.763944   
min      10.000000   0.000000   1.000000   0.750000   0.000000   0.000000   
25%     160.500000   0.000000   2.000000   3.000000   0.000000   1.000000   
50%     447.000000   1.000000   3.000000   6.000000   1.000000   1.000000   
75%     644.000000   1.000000   3.000000  12.000000   2.000000   2.000000   
max     876.000000   1.000000   3.000000  15.000000   4.000000   2.000000   

             Fare  
count   43.000000  
mean    31.302328  
std     38.369737  
min      7.225000  
25%     14.454200  
50%     22.025000  
75%     31.275000  
max    211.337500

In [ ]:
df_male = df.loc[df.Sex == 'male']
df_male.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   577.000000  577.000000  577.000000  453.000000  577.000000   
mean    454.147314    0.188908    2.389948   30.726645    0.429809   
std     257.486139    0.391775    0.813580   14.678201    1.061811   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     222.000000    0.000000    2.000000   21.000000    0.000000   
50%     464.000000    0.000000    3.000000   29.000000    0.000000   
75%     680.000000    0.000000    3.000000   39.000000    0.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  577.000000  577.000000  
mean     0.235702   25.523893  
std      0.612294   43.138263  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   10.500000  
75%      0.000000   26.550000  
max      5.000000  512.329200

In [ ]:
df_male = df.loc[(df.Sex == 'male') & (df.Age < 16)]
df_male.describe()

PassengerId   Survived     Pclass        Age      SibSp      Parch  \
count    40.000000  40.000000  40.000000  40.000000  40.000000  40.000000   
mean    448.550000   0.525000   2.625000   5.466750   2.225000   1.350000   
std     296.365328   0.505736   0.627878   4.398238   1.775727   0.579567   
min       8.000000   0.000000   1.000000   0.420000   0.000000   0.000000   
25%     180.250000   0.000000   2.000000   1.750000   1.000000   1.000000   
50%     397.500000   1.000000   3.000000   4.000000   1.000000   1.000000   
75%     764.000000   1.000000   3.000000   9.000000   4.000000   2.000000   
max     870.000000   1.000000   3.000000  15.000000   5.000000   2.000000   

             Fare  
count   40.000000  
mean    34.380522  
std     27.647000  
min      7.229200  
25%     18.778125  
50%     29.125000  
75%     39.687500  
max    151.550000

In [ ]:
df_male = df.loc[(df.Sex == 'male') & (df.Age > 25)]
df_male.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   274.000000  274.000000  274.000000  274.000000  274.000000   
mean    458.744526    0.204380    2.135036   39.408759    0.259124   
std     252.499564    0.403985    0.868673   11.468099    0.493927   
min       5.000000    0.000000    1.000000   26.000000    0.000000   
25%     232.250000    0.000000    1.000000   30.000000    0.000000   
50%     465.000000    0.000000    2.000000   36.000000    0.000000   
75%     671.250000    0.000000    3.000000   46.000000    0.000000   
max     891.000000    1.000000    3.000000   80.000000    2.000000   

            Parch        Fare  
count  274.000000  274.000000  
mean     0.178832   29.587392  
std      0.599869   51.574427  
min      0.000000    0.000000  
25%      0.000000    7.925000  
50%      0.000000   13.250000  
75%      0.000000   30.000000  
max      5.000000  512.329200

##2. Operaciones Básicas con DataFrames



###2.1. Acceso a datos

**Conceptos clave:**

- Acceso a columnas y filas.
- Uso de loc (basado en etiquetas) y iloc (basado en posiciones).


**Ejemplo:**

In [ ]:
import pandas as pd

# Crear un DataFrame de ejemplo
data = {
    'Nombre': ['Ana', 'Luis', 'Carlos', 'María'],
    'Edad': [23, 45, 34, 29],
    'Ciudad': ['Madrid', 'Barcelona', 'Valencia', 'Sevilla']
}

df = pd.DataFrame(data)

In [ ]:
# Acceder a una columna
edades = df['Edad']
print("Edades:")
print(edades)

In [ ]:

# Acceder a múltiples columnas
nombre_ciudad = df[['Nombre', 'Ciudad']]
print("\nNombre y Ciudad:")
print(nombre_ciudad)

In [ ]:
# Acceder a una fila por etiqueta
fila_ana = df.loc[0]
print("\nFila de Ana:")
print(fila_ana)

In [ ]:
# Acceder a una fila por posición
fila_primera = df.iloc[0]
print("\nPrimera fila:")
print(fila_primera)


### 2.2. Selección y filtrado

Conceptos clave:

- Filtrado condicional.
- Uso de operadores lógicos (&, |, ~).

**Ejemplo:**

In [ ]:
# Filtrar por condición
mayores_30 = df[df['Edad'] > 30]
print("\nPersonas mayores de 30 años:")
print(mayores_30)

In [ ]:
# Filtrar por múltiples condiciones
madrid_mayor_25 = df[(df['Ciudad'] == 'Madrid') & (df['Edad'] > 25)]
print("\nPersonas de Madrid mayores de 25 años:")
print(madrid_mayor_25)

In [ ]:
# Uso de isin para múltiples valores
ciudades = df[df['Ciudad'].isin(['Madrid', 'Valencia'])]
print("\nPersonas de Madrid o Valencia:")
print(ciudades)

### 2.3. Ordenamiento de datos

**Conceptos clave:**

- Ordenar por una o varias columnas.
- Orden ascendente (ascending=True) y descendente (ascending=False).

**Ejemplo:**

In [ ]:
# Ordenar por edad ascendente
df_ordenado_edad = df.sort_values(by='Edad')
print("\nDataFrame ordenado por edad (ascendente):")
print(df_ordenado_edad)


In [ ]:

# Ordenar por edad descendente
df_ordenado_edad_desc = df.sort_values(by='Edad', ascending=False)
print("\nDataFrame ordenado por edad (descendente):")
print(df_ordenado_edad_desc)


In [ ]:

# Ordenar por ciudad y luego por edad
df_ordenado_ciudad_edad = df.sort_values(by=['Ciudad', 'Edad'])
print("\nDataFrame ordenado por ciudad y edad:")
print(df_ordenado_ciudad_edad)


### 2.4. Modificación y eliminación de datos

**Conceptos clave:**

- Añadir y modificar columnas.
- Eliminar filas y columnas.

**Ejemplo:**

In [ ]:
# Añadir una nueva columna
df['Puntaje'] = [88, 92, 79, 85]
print("\nDataFrame con nueva columna 'Puntaje':")
print(df)


In [ ]:
# Modificar valores
df.loc[df['Nombre'] == 'Ana', 'Edad'] = 24
print("\nDataFrame después de modificar la edad de Ana:")
print(df)


In [ ]:
# Eliminar una columna
df_sin_puntaje = df.drop('Puntaje', axis=1)
print("\nDataFrame sin la columna 'Puntaje':")
print(df_sin_puntaje)


In [ ]:
# Eliminar una fila
df_sin_fila = df.drop(2)  # Eliminar la fila con índice 2
print("\nDataFrame sin la fila de Carlos:")
print(df_sin_fila)

### 2.5. Manejo de datos faltantes

**Conceptos clave:**

- Identificar y tratar valores nulos (NaN).
- Métodos para eliminar o imputar datos faltantes.

**Ejemplo:**

In [ ]:
# Introducir valores faltantes
df_con_nan = df.copy()
df_con_nan.loc[1, 'Edad'] = None
print("\nDataFrame con valor faltante:")
print(df_con_nan)


In [ ]:
# Identificar valores faltantes
print("\nValores faltantes en 'Edad':")
print(df_con_nan['Edad'].isnull())

In [ ]:

# Eliminar filas con valores faltantes
df_sin_nan = df_con_nan.dropna()
print("\nDataFrame sin filas con valores faltantes:")
print(df_sin_nan)

In [ ]:
# Imputar valores faltantes
df_imputado = df_con_nan.fillna({'Edad': df_con_nan['Edad'].mean()})
print("\nDataFrame con valores faltantes imputados:")
print(df_imputado)


###2.6. Tamaño y resumen de datos

**Conceptos clave:**

- Obtener el número de filas y columnas.
- Resumen estadístico de los datos.

**Ejemplo:**

In [ ]:
# Obtener el número de filas y columnas
filas, columnas = df.shape
print(f"\nEl DataFrame tiene {filas} filas y {columnas} columnas.")


In [ ]:

# Listar las columnas
print("\nColumnas del DataFrame:")
print(df.columns)

In [ ]:
# Resumen estadístico
print("\nResumen estadístico:")
print(df.describe())

### 2.7. Ejercicios



**Ejercicio 1**

Crea un DataFrame con la siguiente información:

- Nombre: 'Juan', 'Laura', 'Pedro', 'Carmen', 'Luis'
- Departamento: 'Ventas', 'Marketing', 'Ventas', 'Finanzas', 'Marketing'
- Salario: 50000, 60000, 55000, 65000, 62000

Tareas:

- Accede a la columna de salarios.
- Filtra los empleados del departamento de 'Marketing'.
- Ordena el DataFrame por salario de manera descendente.
- Añade una nueva columna 'Bono' que sea el 10% del salario.
- Calcula el salario total (salario + bono) y añade una columna 'SalarioTotal'.


**Ejercicio 2**

Utilizando el DataFrame del ejercicio anterior:

- Modifica el salario de 'Juan' a 52000.
- Elimina el empleado 'Pedro' del DataFrame.
- Identifica si hay datos faltantes en el DataFrame.
- Si hay datos faltantes, imputa con el valor promedio de la columna correspondiente.




**Ejercicio 3**

Crea un DataFrame con datos de ventas:

- Producto: 'Producto A', 'Producto B', 'Producto C', 'Producto D'
- Ventas Enero: 150, 200, 140, 170
- Ventas Febrero: 180, None, 160, 190

Tareas:

- Calcula el total de ventas por producto sumando 'Ventas Enero' y 'Ventas Febrero'.
- Identifica los productos con datos faltantes en 'Ventas Febrero'.
- Imputa los valores faltantes con el promedio de 'Ventas Febrero'.
- Ordena el DataFrame por total de ventas de manera ascendente.

---

# Gracias por completar este laboratorio!

---